File: codes/cpp/polymorphism/subtype.cpp
#include <iostream>
#include <string>

class Person {
public:
    void talk(){
        std::cout << "I am a person" << std::endl;
    }
};

class Student: public Person {};

class Professor: public Person {};

void announce(std::string msg, Person p){
    std::cout << msg << ":";
    p.talk();
}

int main()
{
    Person pe; Student st; Professor pr;

    announce("Person", pe);
    announce("Student", st);
    announce("Professor", pr);

    std::cout << "Table of Person" << std::endl;
    Person people[] = {pe, st, pr};
    for (int i =0; i<3; ++i) { people[i].talk(); }

    return 0;
}

File: codes/java/src/polymorphism/Subtype.java
package polymorphism;

public class Subtype {
    public static void announce(String msg, Person p) {
        System.out.print(msg + ": ");
        p.talk();
    }
    //...
    public static void announce2(String msg, Machine m) {
        System.out.print(msg + ": ");
        m.talk();
    }
    public static void main(String[] args) {
        Person pe = new Person();
        Student st = new Student();
        Professor pr = new Professor();

        announce("Person", pe);
        announce("Student", st);
        announce("Professor", pe);

        System.out.println("Table of Person");
        Person[] people = {pe, st, pr};
        for (Person p: people) { p.talk(); }

        Robot ro = new Robot();
        Cyborg cy = new Cyborg();

        announce2("Robot", ro);
        announce2("Cyborg", cy);
        announce("Cyborg", cy);
    }
}

class Person {
    public void talk(){
        System.out.println("I am a person");
    }
}

interface Machine {
    public void talk();
}

class Student extends Person {}

class Professor extends Person {}

class Robot implements Machine {
    public void talk(){
        System.out.println("I am a robot");
    }
}

class Cyborg extends Person implements Machine {}
